crates/core/component/shielded-pool/src/component/assets.rs

```diff
@@ -0,0 +1,36 @@
+use async_trait::async_trait;
+use cnidarium::{StateRead, StateWrite};
+use penumbra_asset::asset::{self, Metadata};
+use penumbra_proto::{StateReadProto, StateWriteProto};
+
+use tracing::instrument;
+
+use crate::state_key;
+
+#[async_trait]
+pub trait AssetRegistryRead: StateRead {
+    async fn denom_by_asset(&self, asset_id: &asset::Id) -> Option<Metadata> {
+        self.get(&state_key::denom_by_asset(asset_id))
+            .await
+            .expect("no deserialization error")
```

Comment:
Q: Will this potentially fail for benign resigns we nevertheless want to throw upstream, e.g. wrong storage path? IDK what the convention here is though.

Reply:
I don't think so, the failure mode for this is if we have a cache miss that hits storage and fails. The most likely reason this happens is some disk failure that cause storage degradation. Either way, it's irrecoverable for callers and the node should halt.